Utilize EpochNanoseconds

JohannesHelleve opened the pull request:
Solves #119 
Started fresh up the code, one test failing in the instant class still. 

Im a commits behind, but wanted to merge when the code was working locally first.

Some things i'm unsure about:
- When to use borrowed or owned. Used both in different places, but not sure what is optimal.
- When to use EpochNanoseconds(...) or EpochNanoseconds::try_from(...)?/.unwrap

Thanks!

Let me know if you have any quick comments

nekevss commented on src/builtins/core/instant.rs:
```diff
@@ -157,16 +157,16 @@ impl Instant {
 
     /// Creates a new `Instant` from the provided Epoch millisecond value.
-    pub fn from_epoch_milliseconds(epoch_milliseconds: i64) -> TemporalResult<Self> {
+    pub fn from_epoch_milliseconds(epoch_milliseconds: EpochNanoseconds) -> TemporalResult<Self> {
```
Hmmmmm, I believe this isn't entirely correct since we are dealing with milliseconds vs. nanoseconds. This should probably be left as is or a new `EpochMilliseconds` primitive type could be introduces with an internal representation of `i64`.

nekevss commented on src/builtins/core/timezone.rs:
```diff
@@ -140,24 +142,24 @@ impl TimeZone {
-        let nanos = self.get_offset_nanos_for(instant.as_i128(), provider)?;
-        IsoDateTime::from_epoch_nanos(instant.epoch_nanoseconds(), nanos.to_i64().unwrap_or(0))
+        let nanos = self.get_offset_nanos_for(instant.epoch_nanoseconds(), provider)?;
+        IsoDateTime::from_epoch_nanos(instant.epoch_nanoseconds(), nanos.as_i128().to_i64().unwrap_or(0))
```
thought: I think `to_i64` could be implemented directly on `EpochNanoseconds` as a method. We're already doing the bound check, so it could be a nice addition from an ergonomics perspective.

nekevss reviewed (commented):
Was taking a look through, and thought I'd leave a couple comments / thoughts that I noticed.

JohannesHelleve commented:
Not so well with git so not sure to do with that mistake. 

Updated from your feedback. Thanks!

Added the helper method.

Altered a test, but should have the same functionality?

nekevss reviewed (commented):
Hey! Sorry for the delayed review on this. To be honest, I'm not entirely sure what to make of this PR at least until we have built out the provider crate a bit more thoroughly. I'm fairly convinced that using `EpochNanoseconds` more thoroughly is a good thing, but I'm not entirely sure to what degree currently.

I think this may be soft blocked until we are a bit further into the process of setting up the provider crate and have a better idea of what should live where.

I left some comments explaining my general thoughts.

Manishearth commented:
Note: https://github.com/boa-dev/temporal/pull/452 changes how we look at this type

nekevss commented:
Closing this PR. We've moved to a bit of a different way of approaching this type. And as addressed by #452, intermediate nanoseconds can be outside of a valid range, which undercuts the initial purpose of expanding EpochNanosecond's use.
